lib/charms/vault_k8s/v0/juju_facade.py

```diff
@@ -0,0 +1,522 @@
+    def _read_relation_data(
+        self, relation_name: str, relation_id: int, entity: Unit | Application
+    ) -> RelationDataContent | dict[str, str]:
+        relation = self.get_relation_by_id(relation_name, relation_id)
+        if not relation:
+            raise NoSuchRelationError(f"Relation {relation_name}:{relation_id} not found")
+        return relation.data.get(entity, {})
+
+    def get_app_relation_data(
+        self, relation_name: str, relation_id: int
+    ) -> RelationDataContent | dict[str, str]:
+        """Get relation data from the caller's application databag.
+
+        Returns:
+            The relation data as a dict
+
+        Raises:
+            NoSuchRelationError: if the relation is not found
+        """
+        try:
+            return self._read_relation_data(relation_name, relation_id, self.charm.model.app)
+        except NoSuchRelationError:
+            raise
+
+    def get_remote_app_relation_data(
+        self, relation_name: str, relation_id: int
+    ) -> RelationDataContent | dict[str, str]:
+        """Get relation data from the remote application databag.
+
+        Returns:
+            The relation data as a dict
+
+        Raises:
+            NoSuchRelationError: if the relation is not found
+        """
+        try:
+            return self._read_relation_data(relation_name, relation_id, self.charm.model.app)
+        except NoSuchRelationError:
+            raise
+
+    def get_unit_relation_data(
+        self, relation_name: str, relation_id: int
+    ) -> RelationDataContent | dict[str, str]:
+        """Get relation data from the remote unit databag.
+
+        Raises:
+            NoSuchRelationError: if the relation is not found
+        """
+        try:
+            return self._read_relation_data(relation_name, relation_id, self.charm.model.unit)
+        except NoSuchRelationError:
+            raise
+
+    def get_remote_units_relation_data(
+        self, relation_name: str, relation_id: int
+    ) -> List[RelationDataContent | dict[str, str]]:
+        """Get relation data from the remote units databags.
+
+        Raises:
+            NoSuchRelationError: if the relation is not found
+        """
+        relation = self.get_relation_by_id(relation_name, relation_id)
+        if not relation:
+            raise NoSuchRelationError(f"Relation {relation_name}:{relation_id} not found")
+        return [relation.data.get(unit, {}) for unit in relation.units]
+
+    def _set_relation_data(
+        self,
+        data: dict[str, str],
+        relation_name: str,
+        relation_id: int,
+        entity: Unit | Application,
+    ) -> None:
+        relation = self.get_relation_by_id(relation_name, relation_id)
+        if not relation:
+            raise NoSuchRelationError(f"Relation {relation_name}:{relation_id} not found")
+        if not all(isinstance(value, str) for value in data.values()) or not all(
+            isinstance(key, str) for key in data.keys()
+        ):
```

Comment:
This could be a bit less verbose by using `chain()` from `itertools`:

```suggestion
from itertools import chain
...

        if not all(isinstance(value, str) for value in chain(data.values(), data.keys())):
```

This is just suggestion, I would not block the PR for this.